File: backend/app/models/trainer.py
"""Trainer profile — the role-specific data for a User whose role is `trainer`.

Mirrors the fields the admin portal's mock `trainers` array already uses
(specialization, status, capacity, ...) so the frontend can switch over with
minimal change.
"""

import enum
import uuid

from sqlmodel import Field

from app.models.base import TimestampedBase


class TrainerStatus(str, enum.Enum):
    active = "active"
    inactive = "inactive"


class Trainer(TimestampedBase, table=True):
    __tablename__ = "trainers"

    user_id: uuid.UUID = Field(foreign_key="users.id", unique=True, index=True)
    specialization: str
    status: TrainerStatus = Field(default=TrainerStatus.active, index=True)
    capacity: int = Field(default=20, ge=0)
    # One immutable referral code per trainer (see data/referrals.json).
    referral_code: str | None = Field(default=None, unique=True, index=True)
    # The code this trainer signed up with, if any (full referral tracking is a
    # separate feature — this just records what was entered at registration).
    referred_by_code: str | None = Field(default=None, index=True)
    bio: str | None = None

File: backend/src/models/Trainer.js
import mongoose from 'mongoose'
import { ACCOUNT_STATUS } from '../config/constants.js'

// Trainer profile. One-to-one with a User (role: 'trainer').
// Fields mirror admin/src/services/mockData.js `trainers[]` plus the referral
// registry (data/referrals.json) surfaced in trainer Settings + admin Referrals.
const trainerSchema = new mongoose.Schema(
    {
        user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, unique: true },

        specialization: { type: String, default: 'General Fitness', trim: true },

        // "18 / 25" capacity widget (admin CapacityBar). `clientCount` is a
        // denormalised counter kept in step with Client.trainer assignments; the
        // authoritative number is always Client.countDocuments({ trainer }).
        capacity: { type: Number, default: 20, min: 0 },
        clientCount: { type: Number, default: 0, min: 0 },

        rating: { type: Number, default: 5, min: 0, max: 5 },
        revenue: { type: Number, default: 0, min: 0 }, // lifetime, USD
        status: { type: String, enum: ACCOUNT_STATUS, default: 'active', index: true },
        joinDate: { type: Date, default: Date.now },

        // Trainer-to-trainer referrals. Code is issued ONCE and is immutable
        // (trainer/src/services/referrals.js). `referredBy` is a one-time link.
        referralCode: { type: String, unique: true, sparse: true, uppercase: true, trim: true },
        referredBy: { type: mongoose.Schema.Types.ObjectId, ref: 'Trainer', default: null },
    },
    { timestamps: true },
)

trainerSchema.virtual('availableSlots').get(function () {
    return Math.max(0, (this.capacity || 0) - (this.clientCount || 0))
})

trainerSchema.set('toJSON', { virtuals: true })

export const Trainer = mongoose.model('Trainer', trainerSchema)
